### PR TITLE
ci: Adding a SonarQube config file to match Unreal Engine development

### PR DESCRIPTION
**Pull request by baxeaz**
### What was the problem/requirement? (What/Why)

Unreal has a number of coding conventions which don't match default C++. Some of these were being flagged by SonarQube as incorrect. This will lead to unnecessary development churn and frustration.

### What was the solution? (How)

Add a custom sonarqube configuration file to allow Unreal conventions.

### What is the impact of this change?

Better developer experience.

### How was this change tested?

Testing with an incoming PR.

### Was this change documented?

No

### Is this a breaking change?

No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: sonar-project.properties [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+# Exclude build directories and third-party code
+sonar.exclusions=**/Build/**,**/Intermediate/**,**/ThirdParty/**,**/Content/**
+
+# Configure C++ analysis
+sonar.cpp.errorRecoveryEnabled=true
+sonar.cpp.microsoft=true
+sonar.cpp.standardLibrary=microsoft
+
+# Ignore Unreal-specific naming conventions
+sonar.issue.ignore.multicriteria=e1,e2,e3,e4,e5,e6
```
Comment (mwiebe) [inline sonar-project.properties]: What does this specifically do, how does it relate to naming conventions?

File: sonar-project.properties [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+# Custom parameters for Unreal-specific patterns
+sonar.cplusplus.complexity.threshold=15
+sonar.cplusplus.lines.of.code.per.file.threshold=1000
+sonar.cplusplus.maximum.function.parameters.threshold=8
+
+# Allow UPROPERTY macro usage
```
Comment (eherozhao) [inline sonar-project.properties]: May I ask why do we choose these rule key? Any guideline we followed?

File: sonar-project.properties [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+sonar.issue.ignore.multicriteria.e2.resourceKey=**/*.h
+
+# Allow enum "Collisions" for enum BlueprintType declarations
+sonar.issue.ignore.multicriteria.e2.ruleKey=cpp:S841
+sonar.issue.ignore.multicriteria.e2.resourceKey=**/*.h
+
+# Enable specific rules useful for Unreal development
+sonar.cpp.suffixes.headers=.h,.hpp
+sonar.cpp.suffixes.sources=.cpp
+sonar.c.file.suffixes=
```
Comment (mwiebe) [inline sonar-project.properties]: Could these go in the top block along with `sonar.cpp.file.suffixes=.cpp,.h,.hpp`? They look thematically similar to those.

File: sonar-project.properties (hunk 2 above)
Comment (baxeaz) [reply sonar-project.properties]: It's a good call out - I really don't think we need these at all.  Removed this and several other parts I'm not yet sure we need.

File: sonar-project.properties (hunk 1 above)
Comment (baxeaz) [reply sonar-project.properties]: Specifically these are just declaring the existence of named ignores we'll use below I believe.  The documentation for this file is not great - a lot of what's here it seems other folks had to just figure out.  SonarQube seems to vastly prefer these are configured through the web UI.  I probably should have asked if we had a policy around this and also preferred setting these in the UI - setting it in a config file didn't seem at all out of the ordinary at first, and likely far more maintainable.

File: sonar-project.properties (hunk 3 above)
Comment (baxeaz) [reply sonar-project.properties]: Actually I'm not sure we need both - Trimmed out a bunch, but going to do some experimenting locally, so not looking for another review quite yet

**Pull request closed without merging**